File: backend/models/patrol_log.py
from dataclasses import dataclass
from typing import Optional


@dataclass
class PatrolLogRecord:
    log_id: str
    guard_id: str
    checkpoint_id: str
    checkpoint_name: str
    expected_arrival: str     # ISO 8601
    actual_arrival: Optional[str]  # ISO 8601; None if guard missed the checkpoint
    dwell_time_seconds: int
    min_dwell_required: int
    ble_detected: bool
    vigi_detected: bool
    compliant: bool
    shift_id: str
    cycle_id: str = ""  # groups checkpoint visits into one real-time patrol lap;
                        # "" for pre-existing/simulated logs that predate this field
    # True only for a checkpoint the time-boxed window closed before reaching
    # (Prompt 122) — distinct from a genuine skip (actual_arrival is None with
    # this False, unchanged meaning). Absent/False on every log written before
    # Prompt 122 — those keep rendering under the old wrap-model semantics,
    # by construction, since this field didn't exist yet.
    not_in_window: bool = False

File: frontend/types/patrolLog.ts
export interface PatrolLogRecord {
  log_id: string;
  guard_id: string;
  checkpoint_id: string;
  checkpoint_name: string;
  expected_arrival: string;       // ISO 8601
  actual_arrival: string | null;  // ISO 8601; null if guard missed checkpoint
  dwell_time_seconds: number;
  min_dwell_required: number;
  ble_detected: boolean;
  vigi_detected: boolean;
  compliant: boolean;
  shift_id: string;
  cycle_id?: string;  // absent on logs predating the real-time tracker (Prompt 110)
  // True only for a checkpoint the time-boxed window closed before reaching
  // (Prompt 122) — never a violation. Absent (falsy) on every log written
  // before Prompt 122, which is exactly right: those render under the old
  // wrap-model semantics unchanged, since the field didn't exist yet.
  not_in_window?: boolean;
}
